core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala

```diff
@@ -70,230 +58,191 @@ class BrokerMetadataListener(
+    // TODO: cache leaderAndEpoch so we can use the epoch in broker-initiated snapshots.
   }
 
-  def handleUnfenceBrokerRecord(imageBuilder: MetadataImageBuilder,
-                                record: UnfenceBrokerRecord): Unit = {
-    // TODO: add broker epoch to metadata cache, and check it here.
-    imageBuilder.brokersBuilder().changeFencing(record.id(), fenced = false)
+  override def beginShutdown(): Unit = {
+    eventQueue.beginShutdown("beginShutdown", new ShutdownEvent())
   }
 
-  def handleRemoveTopicRecord(imageBuilder: MetadataImageBuilder,
-                              record: RemoveTopicRecord): Unit = {
-    imageBuilder.topicIdToName(record.topicId()) match {
-      case None =>
-        throw new RuntimeException(s"Unable to locate topic with ID ${record.topicId}")
-
-      case Some(topicName) =>
-        info(s"Processing deletion of topic $topicName with id ${record.topicId}")
-        val removedPartitions = imageBuilder.partitionsBuilder().removeTopicById(record.topicId())
-        groupCoordinator.handleDeletedPartitions(removedPartitions.map(_.toTopicPartition).toSeq, RequestLocal.NoCaching)
-        configRepository.remove(new ConfigResource(ConfigResource.Type.TOPIC, topicName))
+  class ShutdownEvent() extends EventQueue.FailureLoggingEvent(log) {
+    override def run(): Unit = {
+      removeMetric(BrokerMetadataListener.MetadataBatchProcessingTimeUs)
+      removeMetric(BrokerMetadataListener.MetadataBatchSizes)
     }
   }
 
-  def handleClientQuotaRecord(imageBuilder: MetadataImageBuilder,
-                        record: ClientQuotaRecord): Unit = {
-    // TODO add quotas to MetadataImageBuilder
-    clientQuotaManager.handleQuotaRecord(record)
+  def close(): Unit = {
+    beginShutdown()
+    eventQueue.close()
   }
 
-  def handleProducerIdRecord(record: ProducerIdsRecord): Unit = {
-    // This is a no-op since brokers get their producer ID blocks directly from the controller via
-    // AllocateProducerIds RPC response
+  // VisibleForTesting
+  private[kafka] def getImageRecords(): CompletableFuture[util.List[ApiMessageAndVersion]] = {
+    val future = new CompletableFuture[util.List[ApiMessageAndVersion]]()
+    eventQueue.append(new GetImageRecordsEvent(future))
+    future
   }
 
-  class HandleNewLeaderEvent(leaderAndEpoch: LeaderAndEpoch)
-      extends EventQueue.FailureLoggingEvent(log) {
-    override def run(): Unit = {
-      val imageBuilder =
-        MetadataImageBuilder(brokerId, log, metadataCache.currentImage())
-      imageBuilder.controllerId(leaderAndEpoch.leaderId.asScala)
-      metadataCache.image(imageBuilder.build())
+  class GetImageRecordsEvent(future: CompletableFuture[util.List[ApiMessageAndVersion]])
+      extends EventQueue.FailureLoggingEvent(log) with Consumer[util.List[ApiMessageAndVersion]] {
+    val records = new util.ArrayList[ApiMessageAndVersion]()
+    override def accept(batch: util.List[ApiMessageAndVersion]): Unit = {
+      if (batch == null) {
```

Comment:
Why use the `null` here to finalize the image? Could we complete the future below where we call `_image.write`?

Alternatively, we could use a composite Consumer + Closable type so MetadataImage could call `close` instead of `out.accept(null)`.

Reply:
You're right, there is no need for the null here. Removed.